components/application-authenticators/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/AuthenticationHandler.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2016, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.identity.application.authenticator.iwa;
+
+import org.ietf.jgss.GSSContext;
+import org.ietf.jgss.GSSCredential;
+import org.ietf.jgss.GSSException;
+import org.wso2.carbon.user.core.service.RealmService;
+
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.kerberos.KerberosPrincipal;
+import javax.security.auth.login.LoginContext;
+import javax.security.auth.login.LoginException;
+import java.security.PrivilegedActionException;
+
+/**
+ * this class process the kerberos token and create server credentials
+ */
+public class AuthenticationHandler {
+
+    private static transient GSSCredential serverCred;
+    private static transient KerberosPrincipal serverPrincipal;
+
+    public static void initialize() throws GSSException, PrivilegedActionException, LoginException {
+
+        //todo if this not given fall back to default user name
+        RealmService realmService = IWAServiceDataHolder.getRealmService();
+        String username = realmService.getBootstrapRealmConfiguration().getUserStoreProperty(IWAConstants.SPN_NAME);
+        String password = realmService.getBootstrapRealmConfiguration().getUserStoreProperty(IWAConstants.SPN_PASSWORD);
+        CallbackHandler callbackHandler = IWAServiceDataHolder.getUsernamePasswordHandler(username, password);
+        serviceLogin(callbackHandler);
+        serverPrincipal = new KerberosPrincipal(serverCred.getName().toString());
+    }
+
+    /**
+     * Process kerberos token and get user name
+     *
+     * @param gssToken kerberos token
+     * @return username
+     * @throws GSSException
+     */
+    public static String processToken(byte[] gssToken) throws GSSException {

```

Comment:
Why all the methods are static? May be better to consider making this class singleton.